src/vos/vos_dtx.c

```diff
@@ -108,10 +108,26 @@ dtx_set_aborted(uint32_t *tx_lid)
 }
 
 static inline int
-dtx_inprogress(struct vos_dtx_act_ent *dae, int pos)
+dtx_inprogress(struct vos_dtx_act_ent *dae, bool for_read)
 {
-	D_DEBUG(DB_IO, "Hit uncommitted DTX "DF_DTI" lid=%d at %d\n",
-		DP_DTI(&DAE_XID(dae)), DAE_LID(dae), pos);
+	D_DEBUG(DB_IO, "Hit uncommitted DTX "DF_DTI" lid=%d for %s\n",
+		DP_DTI(&DAE_XID(dae)), DAE_LID(dae),
+		for_read ? "read" : "modification");
+
+	/* If the modifications corsses multiple redundancy groups, then it
+	 * is possible that the sub modifications on the DTX leader are not
+	 * the same as the ones on non-leaders. Under such case, if someone
+	 * wants to read the data on some non-leader but hits non-committed
+	 * DTX, then asking the client to retry with leader maybe not help.
+	 * Instead, we can ask make the client to retry the read again (and
+	 * again) sometime later. We do sync commit the DTX with 'DTE_BLOCK'
+	 * flag. So it will not cause the client to retry read for too many
+	 * times unless such DTX hit some trouble (such as client or server
+	 * failure) that may cause current readers to be blocked until such
+	 * DTX has been handled by the new leader via DTX recovery.
+	 */
+	if (DAE_FLAGS(dae) & DTE_BLOCK && for_read)
+		return -DER_AGAIN;
```

Comment:
so -DER_AGAIN means retry from same server whereas DER_INPROGRESS means go to leader?  I see a few other cases in DTX code where we return -DER_AGAIN.  Is it possible to get the confused?

Reply:
OK, I will define a new errno -DER_TX_BUSY for that.